### PR TITLE
Lorenz ODE

### PR DESCRIPTION
**Pull request by NAThompson**
This example is in response to a conversation we have on teams with @pnorbert .

The example solves the [Lorenz system](https://en.wikipedia.org/wiki/Lorenz_system) via an adaptive Taylor method. It presents a few challenges for ADIOS: First, the number of state points in the state vector varies wildly. Second, the natural datastructure is an array of structs.

The code validates that the array of structs has contiguous layout, as there is nothing [guaranteed by the standard](https://stackoverflow.com/questions/54197195/what-is-the-memory-layout-of-vector-of-arrays). Then it pipes variable-sized data to independent variables, which, though inelegant, is what I understand to be required. (Is there a more elegant way to do it?)

Finally, once I get a little feedback, I'll bring in MPI, and we can really fill up the phase space with trajectories; which should be a very useful simplified model for particle advection.

### CODE REVIEW

File: source/cpp/lorenz_ode/lorenz_writer.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                auto solution = lorenz<double>(sigma, beta, rho, initial_conditions, tmax, absolute_error);
+                auto const & skeleton = solution.state();
+                std::cout << skeleton.size() << " points in solution skeleton.\n";
+                for (size_t i = 0; i < skeleton.size(); ++i) {
+                    const double * const p = skeleton[0].data();
+                    for (size_t j = 0; j < 7; ++j) {
+                        assert(skeleton[i][j] == p[7*i+j]);
+                    }
+                }
+                std::string variable = "u" + std::to_string(i) + std::to_string(j) + std::to_string(k);
```
Comment (williamfgc) [inline source/cpp/lorenz_ode/lorenz_writer.cpp]: add `const`

File: source/cpp/lorenz_ode/lorenz_writer.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    io.DefineAttribute<Real>("ρ", rho);
+    io.DefineAttribute<Real>("‖û-u‖", absolute_error);
+    io.DefineAttribute<std::string>("interpretation", "7D array of structs {tᵢ, xᵢ, yᵢ, zᵢ, ẋᵢ, ẏᵢ, żᵢ}");
+    adios2::Engine adios_engine = io.Open("lorenz.bp", adios2::Mode::Write);
+
+    size_t paths = 2;
+    for (size_t i = 0; i < paths; ++i) {
+        for (size_t j = 0; j < paths; ++j) {
+            for (size_t k = 0; k < paths; ++k) {
+                std::array<Real, 3> initial_conditions{Real(i),Real(j),Real(k)};
```
Comment (williamfgc) [inline source/cpp/lorenz_ode/lorenz_writer.cpp]: `const`?

File: source/cpp/lorenz_ode/lorenz.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if (tmax <= 0)
+        {
+            throw std::domain_error("tmax > 0 is required");
+        }
+        if (absolute_error_goal <= std::numeric_limits<Real>::epsilon())
+        {
+            throw std::domain_error("Abolute error goal > eps is required");
+        }
+        Real t = 0;
+        Real x = initial_conditions[0];
```
Comment (williamfgc) [inline source/cpp/lorenz_ode/lorenz.hpp:34]: `const`

**Review by williamfgc — CHANGES_REQUESTED** [review]
@NAThompson overall this is great. Since this is used for tutorials and newcomers, I'd recommend using `const` whenever possible since it improves tremendously code readability (and compile-time safety).

**Comment by williamfgc** [comment]
Also, please rebase `master`. Github is showing an out-of-date issue.

File: source/cpp/lorenz_ode/lorenz_writer.cpp (hunk 2 above)
Comment (NAThompson) [reply source/cpp/lorenz_ode/lorenz_writer.cpp]: Fixed.

File: source/cpp/lorenz_ode/lorenz_writer.cpp (hunk 1 above)
Comment (NAThompson) [reply source/cpp/lorenz_ode/lorenz_writer.cpp]: Fixed.

**Comment by williamfgc** [comment]
@NAThompson great, thanks!

**Comment by NAThompson** [comment]
@williamfgc : Thanks so much for being so helpful here; I hope you won't find it frustrating that I'm going to put in a PR for a reader as well!

(I should've marked the PR as a draft; my bad. But it appears github doesn't allow you to retroactively mark a PR as a draft after opening it.)

**Comment by williamfgc** [comment]
@NAThompson Happy to help and thanks for adding examples. We don't know where users come from and having examples that relate to their tasks helps a lot.

**Pull request merged**